cmd/dagger/client_add.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"dagger.io/dagger"
+	"github.com/dagger/dagger/engine/client"
+	"github.com/dagger/dagger/engine/client/pathutil"
+	"github.com/spf13/cobra"
+)
+
+var (
+	generator string
+	localSDK  bool
+)
+
+func init() {
+	clientAddCmd.Flags().StringVar(&generator, "generator", "", "Generator to use to generate the client")
+	clientAddCmd.Flags().BoolVar(&localSDK, "local-sdk", false, "Use local SDK dependency")
+}
+
+var clientAddCmd = &cobra.Command{
+	Use:     "add [options] [path]",
+	Short:   "Generate a new Dagger client from the Dagger module",
+	Example: "dagger client add --generator=go ./dagger",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return withEngine(cmd.Context(), client.Params{}, func(ctx context.Context, engineClient *client.Client) error {
+			if generator == "" {
+				return fmt.Errorf("generator must set (ts, go, python or custom generator)")
+			}
+
+			// default the output to the current working directory if it doesn't exist yet
+			cwd, err := pathutil.Getwd()
+			if err != nil {
+				return fmt.Errorf("failed to get current working directory: %w", err)
+			}
+
+			outputPath := filepath.Join(cwd, "dagger")
+			if len(args) > 0 {
+				outputPath = args[0]
+			}
+
+			if filepath.IsAbs(outputPath) {
+				outputPath, err = filepath.Rel(cwd, outputPath)
+				if err != nil {
+					return fmt.Errorf("failed to get relative path: %w", err)
+				}
+			}
+
+			handler := &clientAddHandler{
+				dag:        engineClient.Dagger(),
+				cmd:        cmd,
+				outputPath: outputPath,
+			}
+
+			return handler.Run(ctx)
+		})
+	},
+	Annotations: map[string]string{
+		"experimental": "true",
+	},
+}
+
+type clientAddHandler struct {
+	dag        *dagger.Client
+	cmd        *cobra.Command
+	outputPath string
+}
+
+func (c *clientAddHandler) Run(ctx context.Context) (rerr error) {
+	mod, _, err := initializeClientGeneratorModule(ctx, c.dag, ".")
```

Comment:
The fact that we only load `.` for the module means this code will only work if you run it with the source code, right? That might be okay for typescipt/python, but since Go is compiled that's sort of a big/surprising limitation.

I'm totally fine with merging this PR with that limitation but we should fix that in a follow-up I think. For git deps specifically this should be easy, just need to embed the refs of the defs to load in the compiled bin. It's more complicated for local deps, but could figure something out there too.

Reply:
I'm not sure I understand this comment, because there's a find up logic in the engine, I would assume that the whole ctx dir can be accessible?
Do you have an example to share so I can have a better understanding

Reply:
In all your existing examples for `go`, you are running it with `go run main.go`. That will only work if `go run main.go` is run in the same directory that has the module source.

It should be possible to do `go build -o binary main.go`, move `binary` anywhere and then execute `./binary` and everything still works.

The compiled binary just needs to have the list of deps to load embedded in it. Then it should work seamlessly for git deps. Local deps are more complicated like I said but various options possible when we get there.

Reply:
Ohhh I see, yeah you're totally right! the binary needs to have access to the dagger.json and local deps, I'm not sure how this would work